crates/ra_ide/src/syntax_highlighting.rs

```diff
@@ -161,26 +221,43 @@ pub(crate) fn highlight(
 
         if let Some(token) = element.as_token().cloned().and_then(ast::RawString::cast) {
             let expanded = element_to_highlight.as_token().unwrap().clone();
-            if highlight_injection(current, &sema, token, expanded).is_some() {
+            if highlight_injection(&mut stack, &sema, token, expanded).is_some() {
                 continue;
             }
         }
 
+        let is_format_string =
+            format_string.as_ref().map(|fs| fs == &element_to_highlight).unwrap_or_default();
+
         if let Some((highlight, binding_hash)) =
             highlight_element(&sema, &mut bindings_shadow_count, element_to_highlight)
         {
-            current.push(HighlightedRange { range, highlight, binding_hash });
+            stack.add(HighlightedRange { range, highlight, binding_hash });
+            if is_format_string {
+                stack.push();
+                lazy_static! {
+                    static ref RE: Regex =
+                        Regex::new(r"[^\{](?P<format>\{(?:\}|[^\{].*?\}))[^\}]").unwrap();
+                }
+                for (start, end) in RE
```

Comment:
Rather then using ad-hoc regex parsing, I suggest that we install some future proofing here. Eventually, we'd want to handle and color escape sequences properly, as well as check the semantics of format specifiers inside `{}`. 

So I think we need a new function, `lex_string_literal`, which returns inner "tokens" comprising the string. I think it should live in `ra_syntax/ast/tokens.rs` as a method on `ast::String`. Ideally, it should have the following return type

```rust
enum StrngPiece {
    Quote,
    LiteralSequence,
    EscapeSequence,
    FormatSpecifier,
}

fn lex(&self) -> impl Iterator<Item = (TextRange, StringPIece)> { ... }
```

We, however, already use internal iteration for escape sequences (see the call to `unescape_str`), so, instead of returning an iterator, we might just accept the function with `(TextRange, StringPIece)` argument. I don't think we necessary need to handle escape sequences in this PR (although we might do that as well, they also should be colored), but I think it's important to properly isolate string lexer into a separate function, becase we'll have to extend it eventually to handle everything. Oh, and I think it's better to avoid regex here and code everything in a manual way, as we'd need that for escape sequences anyway.

Reply:
Thanks to your detailed description it was straightforward to implement 👍 
I have updated the PR to include the manual lexer (for format specifier and escape sequences). I have also included some tests for the happy path, in case of an error I opted to just return the neutral `StringPiece::LiteralSequence`.
The current implementation does not support RawStrings, though.